File: clients/ios/Lextures/Core/Config/AppConfiguration.swift
import Foundation

/// Runtime API configuration. Debug builds read `API_BASE_URL` from `Config/Development.xcconfig` (see `clients/scripts/setup-mobile-dev.sh`).
enum AppConfiguration {
    private static let defaultAPIBase = "http://127.0.0.1:8080"

    static var apiBaseURL: URL {
        if let env = ProcessInfo.processInfo.environment["API_BASE_URL"],
           let url = URL(string: env), !env.isEmpty {
            return url
        }
        if let plist = Bundle.main.object(forInfoDictionaryKey: "API_BASE_URL") as? String,
           !plist.isEmpty,
           !plist.hasPrefix("$("),
           let url = URL(string: plist) {
            return url
        }
        return URL(string: defaultAPIBase)!
    }

    static func apiURL(path: String) -> URL {
        let base = apiBaseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let normalized = path.hasPrefix("/") ? path : "/\(path)"
        return URL(string: base + normalized)!
    }

    /// Converts the API base (http/https) to a WebSocket URL (ws/wss) for realtime endpoints.
    static func webSocketURL(path: String) -> URL {
        let apiURL = apiURL(path: path)
        var components = URLComponents(url: apiURL, resolvingAgainstBaseURL: false)!
        switch components.scheme {
        case "https": components.scheme = "wss"
        default: components.scheme = "ws"
        }
        return components.url!
    }

    /// Public web pages (privacy/trust center, accessibility statement) are served
    /// from the same origin as the API in this monorepo deployment.
    static func webURL(path: String) -> URL { apiURL(path: path) }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/config/AppConfiguration.kt
package com.lextures.android.core.config

import com.lextures.android.BuildConfig
import java.net.URL

/**
 * Runtime API configuration.
 *
 * Priority:
 * 1. Environment chosen on the get-started screen ([EnvironmentStore])
 * 2. `BuildConfig.API_BASE_URL` from `local.properties` / Gradle `-PAPI_BASE_URL`
 * 3. Emulator loopback default
 */
object AppConfiguration {
    private const val DEFAULT_API_BASE = "http://10.0.2.2:8080"

    /**
     * Optional process-wide override set from [EnvironmentStore] (or tests).
     * Prefer calling [bindEnvironment] once at app start.
     */
    @Volatile
    var runtimeBaseUrl: String? = null

    fun bindEnvironment(store: EnvironmentStore) {
        runtimeBaseUrl = store.apiBaseUrl?.trimEnd('/')
    }

    val apiBaseUrl: String
        get() {
            val runtime = runtimeBaseUrl?.trim().orEmpty()
            if (runtime.isNotEmpty()) {
                return runtime.trimEnd('/')
            }
            val configured = BuildConfig.API_BASE_URL.trim()
            if (configured.isNotEmpty()) {
                return configured.trimEnd('/')
            }
            return DEFAULT_API_BASE
        }

    fun apiUrl(path: String): URL {
        val normalized = if (path.startsWith("/")) path else "/$path"
        return URL(apiBaseUrl + normalized)
    }

    /**
     * Public web pages (privacy/trust center, accessibility statement) are served
     * from the same origin as the API in this monorepo deployment.
     */
    fun webUrl(path: String): String {
        val normalized = if (path.startsWith("/")) path else "/$path"
        return apiBaseUrl + normalized
    }
}
